File: Leetcode Daily Challenge/025-July-2025/031-bitwise-ors-of-subarrays.java
// https://leetcode.com/problems/bitwise-ors-of-subarrays/




import java.util.*;

class Solution {
    // TC : O(N * 32)
    // SC : O(N)
    public int subarrayBitwiseORs(int[] arr) {
        Set<Integer> prev = new HashSet<>();
        Set<Integer> res  = new HashSet<>();

        for (int num : arr) {
            Set<Integer> curr = new HashSet<>();
            for (int val : prev) {
                curr.add(num | val);
                res.add(num | val);
            }
            res.add(num);
            curr.add(num);
            prev = curr;
        }

        return res.size();
    }
}

File: Leetcode Daily Challenge/025-July-2025/031-bitwise-ors-of-subarrays.cpp
// https://leetcode.com/problems/bitwise-ors-of-subarrays/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // TC : O(N)
    // SC : O(N)
    int subarrayBitwiseORs(vector<int>& arr) {
        unordered_set<int> prev, result;
        for (int &num : arr) {
            unordered_set<int> curr;
            for (auto &val : prev) {
                curr.insert(val | num);
                result.insert(val | num);
            }
            curr.insert(num);
            result.insert(num);
            prev = curr;
        }
        return result.size();
    }
};